### PR TITLE
Add note about named pipes on Windows in CLI help and fail on Windows when named pipe has incorrect form

### LINKED ISSUE #2340: Can't start process on Windows 10 - DeserialiseFailure 2 "codecHandshake.Confirm: unexpected key"
# Context
 
Started cardano-node with this:

```
set CARDANO_NODE_SOCKET_PATH=C:\Users\jbax\Downloads\blockchain\node.socket

cardano-node.exe run --topology C:\Users\jbax\dev\free-commerce\src\main\resources\mainnet-topology.json --database-path C:\Users\jbax\Downloads\blockchain --socket-path C:\Users\jbax\Downloads\blockchain\node.socket --host-addr 0.0.0.0 --port 3333 --config C:\Users\jbax\dev\free-commerce\src\main\resources\mainnet-config.json
```

Then cardano-wallet with this:

```
set CARDANO_NODE_SOCKET_PATH=C:\Users\jbax\Downloads\blockchain\node.socket

C:\Users\jbax\Downloads\cardano-wallet-win64>cardano-wallet.exe serve --mainnet --database C:\Users\jbax\Downloads\blockchain --node-socket C:\Users\jbax\Downloads\blockchain\node.socket --port 4444
```

> NOTE: (also tried without setting the `CARDANO_NODE_SOCKET_PATH`)

The cardano-wallet process fails almost immediately with:

```
[cardano-wallet.main:Info:3] [2020-11-22 21:44:50.77 UTC] Running as v2020.11.3 (git revision: e08d0aa5bf28632ec772a7e1b00c421c584dcb35)
[cardano-wallet.main:Info:3] [2020-11-22 21:44:50.77 UTC] Command line: C:\Users\jbax\Downloads\cardano-wallet-win64\cardano-wallet.exe serve --mainnet --database C:\Users\jbax\Downloads\blockchain --node-socket C:\Users\jbax\Downlo
ads\blockchain\node.socket --port 4444
[cardano-wallet.main:Debug:3] [2020-11-22 21:44:50.77 UTC] ServeArgs {_hostPreference = Host "127.0.0.1", _listen = ListenOnPort 4444, _tlsConfig = Nothing, _nodeSocket = "C:\\Users\\jbax\\Downloads\\blockchain\\node.socket", _netwo
rkConfiguration = MainnetConfig, _database = Just "C:\\Users\\jbax\\Downloads\\blockchain", _syncTolerance = SyncTolerance 300s, _enableShutdownHandler = False, _poolMetadataSourceOpt = Nothing, _logging = LoggingOptions {loggingMin
Severity = Debug, loggingTracers = Tracers {applicationTracer = Const (Just Info), apiServerTracer = Const (Just Info), walletEngineTracer = Const (Just Info), walletDbTracer = Const (Just Info), poolsEngineTracer = Const (Just Info
), poolsDbTracer = Const (Just Info), ntpClientTracer = Const (Just Info), networkTracer = Const (Just Info)}, loggingTracersDoc = Nothing}}
[cardano-wallet.main:Info:3] [2020-11-22 21:44:50.77 UTC] Wallet databases: Using directory: C:\Users\jbax\Downloads\blockchain
[cardano-wallet.application:Info:3] [2020-11-22 21:44:50.77 UTC] Wallet backend server starting. Using "C:\\Users\\jbax\\Downloads\\blockchain\\node.socket".
[cardano-wallet.application:Info:3] [2020-11-22 21:44:50.77 UTC] Node is Haskell Node on mainnet.
[cardano-wallet.ntp-client:Info:3] [2020-11-22 21:44:50.79 UTC] Starting ntp client
[cardano-wallet.ntp-client:Info:16] [2020-11-22 21:44:50.79 UTC] query to ntp client invoked
[cardano-wallet.pools-db:Info:3] [2020-11-22 21:44:50.79 UTC] Will open db at C:\Users\jbax\Downloads\blockchain\stake-pools.sqlite
[cardano-wallet.network:Info:10] [2020-11-22 21:44:50.81 UTC] conn:C:\Users\jbax\Downloads\blockchain\node.socket:C:\Users\jbax\Downloads\blockchain\node.socket Send (ClientAgency TokPropose,MsgProposeVersions (fromList [(NodeToClie
ntV_3,TInt 764824073)]))
[cardano-wallet.ntp-client:Notice:21] [2020-11-22 21:44:50.82 UTC] ntp client experienced error withIODataPtr (sendBufTo): invalid argument (The network location cannot be reached. For information about network troubleshooting, see
Windows Help.) when sending packet
[cardano-wallet.main:Debug:3] [2020-11-22 21:44:50.86 UTC] Logging shutdown.
cardano-wallet.exe: ExceptionInLinkedThread ThreadId 10 DecoderFailure (Handshake) ServerAgency TokConfirm) (DeserialiseFailure 2 "codecHandshake.Confirm: unexpected key")
```

If I stop the node and delete the `socket.path` file, the cardano-wallet starts and tries to connect.

Then, if I start the `cardano-node` again, the `cardano-wallet` doesn't connect to the node but keeps retrying.

| Information  | -                                                     |
| ---          | ---                                                   |
| Version      | 2020.11.3 (git revision: e08d0aa5bf28632ec772a7e1b00c421c584dcb35)  |
| Platform     | Windows 10          |
| Installation | From Github Release            |

### PR DESCRIPTION
**Pull request by piotr-iohk**
# Issue Number

#2340 

# Overview

- d597250c57d8fba12c3152889449c500f99fa4d0
  Add note about named pipes on Windows in CLI help for --node-socket
  
- 5437fcdeb7352ebb70f1229b428a59e86e660c20
  Add socket/pipe name validation to --node-socket option
  If the socket/pipe name is invalid, it will say one of:

  option --node-socket: Invalid file path

  option --node-socket: Invalid pipe name. Windows named pipes are of the form \\.\pipe\cardano-node

# Comments

Named pipes are used for communication between cardano-node and cardano-wallet on Windows. As far as I see it wasn't documented anywhere.
Before:
```
cardano-wallet serve
...
  --node-socket FILE       Path to the node's domain socket.
...
```
After:
```
cardano-wallet serve
...
  --node-socket FILE       Path to the node's domain socket file (POSIX) or pipe
                           name (Windows). Note: Maximum length for POSIX socket
                           files is approx. 100 bytes. Note: Windows named pipes
                           are of the form \\.\pipe\cardano-node

...
```

### CODE REVIEW

File: lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs [hunk 1]
```diff
@@ -242,7 +242,9 @@ nodeSocketOption :: Parser FilePath
 nodeSocketOption = optionT $ mempty
     <> long "node-socket"
     <> metavar "FILE"
-    <> help "Path to the node's domain socket."
+    <> help "Path to the node's domain socket. \
+            \Note: named pipes are used on Windows \
+            \e.g. \\\\.\\pipe\\cardano-node"
```
Comment (rvl) [inline lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs]: Maybe enhance it something like this? 

```suggestion
    <> help "Path to the node's domain socket file (POSIX) or pipe name (Windows). \
            \ Note: Maximum length for POSIX socket files is approx. 100 bytes. \
            \ Note: Windows named pipes are of the form \\\\.\\pipe\\cardano-node"
```

**Review by rvl — APPROVED** [review]
Ah, of course! Nice one. I totally forgot about that.

We should update the wiki too.

There are socket filename gotchas on Linux and macOS too.

**Comment by rvl** [comment]
Perhaps we could make the CLI argument parser fail if the `--node-socket` doesn't start with `\\.pipe\` when on windows?

**Comment by piotr-iohk** [comment]
> Perhaps we could make the CLI argument parser fail if the --node-socket doesn't start with \\.pipe\ when on windows?

Pls have look -> 8239e45. :pray: 

> There are socket filename gotchas on Linux and macOS too.

I'm not aware about any, what are those? :sweat_smile:

**Comment by rvl** [comment]
@piotr-iohk I have added a change which checks the socket path/pipe name during CLI option parsing. The error messages are like:

    option --node-socket: Invalid pipe name. Windows named pipes are of the form \\.\pipe\cardano-node

On Linux and macOS the maximum length of socket file paths is 108 and 104 bytes respectively. But I am reluctant to add length validation in case it somehow accidentally excludes valid socket file paths.

File: lib/shelley/test/unit/Cardano/Wallet/Shelley/LaunchSpec.hs [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        let runTest (desc, arg, posix, windows) = it desc $ do
+                let tst = if isWindows then windows else posix
+                parse arg `shouldSatisfy` tst
+        mapM_ runTest
+            [ ( "valid rel path",   "hello.sock",   ok,  err )
+            , ( "valid abs path",   "/run/cw.sock", ok,  err )
+            , ( "invalid",          "",             err, err )
+            , ( "valid pipename",   "\\\\.pipe\\a", ok,  ok  )
+            , ( "invalid pipename", "\\\\.pipe\\",  ok,  err )
+            , ( "windows path",     "c:\\windows",  ok,  err )
```
Comment (piotr-iohk) [inline lib/shelley/test/unit/Cardano/Wallet/Shelley/LaunchSpec.hs]: :heart:

File: lib/shelley/exe/cardano-wallet.hs [hunk 3]
```diff
@@ -223,6 +225,11 @@ cmdServe = command "serve" $ info (helper <*> helper' <*> cmd) $ mempty
                                 exitWith $ ExitFailure 33
 
                 whenJust databaseDir $ setupDirectory (logInfo tr . MsgSetupDatabases)
+                let socket = T.pack nodeSocket
+                when (os == "mingw32" && not ("\\\\.\\pipe\\" `T.isPrefixOf` socket)) $ do
+                    logError tr (MsgInvalidWinNamedPipe socket)
+                    exitWith $ ExitFailure 34
```
Comment (KtorZ) [inline lib/shelley/exe/cardano-wallet.hs]: This validation is now redundant with the validation done in the command-line I believe :thinking:

**Review by KtorZ — APPROVED** [review]
bors merge

**Review by rvl — APPROVED** [review]

**Comment by KtorZ** [comment]
bors merge

**Pull request merged**